src/tools/build_helper/src/git.rs

```diff
@@ -167,29 +167,59 @@ pub fn get_git_untracked_files(
 ///
 /// This can result in formatting thousands of files instead of a dozen,
 /// so we should warn the user something is wrong.
-pub fn warn_old_master_branch(
-    config: &GitConfig<'_>,
-    git_dir: &Path,
-) -> Result<(), Box<dyn std::error::Error>> {
-    use std::time::Duration;
-    const WARN_AFTER: Duration = Duration::from_secs(60 * 60 * 24 * 10);
-    let updated_master = updated_master_branch(config, Some(git_dir))?;
-    let branch_path = git_dir.join(".git/refs/remotes").join(&updated_master);
-    match std::fs::metadata(branch_path) {
-        Ok(meta) => {
-            if meta.modified()?.elapsed()? > WARN_AFTER {
-                eprintln!("warning: {updated_master} has not been updated in 10 days");
-            } else {
-                return Ok(());
+pub fn warn_old_master_branch(config: &GitConfig<'_>, git_dir: &Path) {
+    if crate::ci::CiEnv::is_ci() {
+        // this warning is useless in CI,
+        // and CI probably won't have the right branches anyway.
+        return;
+    }
+    // this will be overwritten by the actual name, if possible
+    let mut updated_master = "the upstream master branch".to_string();
+    match warn_old_master_branch_(config, git_dir, &mut updated_master) {
+        Ok(branch_is_old) => {
+            if !branch_is_old {
+                return;
             }
+            // otherwise fall through and print the rest of the warning
         }
         Err(err) => {
             eprintln!("warning: unable to check if {updated_master} is old due to error: {err}")
         }
     }
     eprintln!(
         "warning: {updated_master} is used to determine if files have been modified\n\
-               warning: if it is not updated, this may cause files to be needlessly reformatted"
+         warning: if it is not updated, this may cause files to be needlessly reformatted"
     );
-    Ok(())
+}
+
+pub fn warn_old_master_branch_(
+    config: &GitConfig<'_>,
+    git_dir: &Path,
+    updated_master: &mut String,
+) -> Result<bool, Box<dyn std::error::Error>> {
+    use std::time::Duration;
+    *updated_master = updated_master_branch(config, Some(git_dir))?;
+    let branch_path = git_dir.join(".git/refs/remotes").join(&updated_master);
+    const WARN_AFTER: Duration = Duration::from_secs(60 * 60 * 24 * 10);
+    let meta = match std::fs::metadata(&branch_path) {
+        Ok(meta) => meta,
+        Err(err) => {
+            let gcd = git_common_dir(&git_dir)?;
```

Comment:
i'm trying not to have much of a performance impact over such a niche lint. i'm stuck on an hdd system for the time being, so `git` can take almost half a second to query basic information if it isn't in the I/O cache.. compared to a single `stat` syscall, which is always nearly instant.

yes, this does give slightly worse performance in the worktree case, but worktrees are already nearly unusable on a slow hdd system, so i'm not too concerned about that.

Reply:
I'm concerned about maintainability of this codebase and how much we're willing to carry difficult-to-test fallback paths in code that is proven to be error-prone and not covered by CI. There's a global cost to every hack like this, and at some point making a slow system a bit slower is an okay way to keep the rest of the project productive -- the time spent figuring out why my worktrees now show this warning is time not spent working on rustc itself, after all.

In the end this will be up to the bootstrap team to decide though, these are just my thoughts.